server/stats.go

```diff
@@ -236,6 +251,84 @@ func (s *Server) GetStatSeries(ctx context.Context, in *pb.GetStatSeriesRequest)
 	return &resp, nil
 }
 
+// GetStatAll implements API for Mixer.GetStatAll.
+func (s *Server) GetStatAll(ctx context.Context, in *pb.GetStatAllRequest) (
+	*pb.GetStatAllResponse, error) {
+	places := in.GetPlaces()
+	statVars := in.GetStatVars()
+	if len(places) == 0 {
+		return nil, fmt.Errorf("Missing required argument: 'places'")
+	}
+	if len(statVars) == 0 {
+		return nil, fmt.Errorf("Missing required argument: 'stat_vars'")
+	}
+	// Read triples for statistical variable.
+	triplesRowList := buildTriplesKey(statVars)
+	triples, err := readTriples(ctx, s.btTable, triplesRowList)
+	if err != nil {
+		return nil, err
+	}
+	statVarObject := map[string]*StatisticalVariable{}
+	for statVar, triplesCache := range triples {
+		if triplesCache != nil {
+			statVarObject[statVar], err = triplesToStatsVar(statVar, triplesCache)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	// Construct BigTable row keys.
+	rowList, keyTokens := buildStatsKey(places, statVarObject)
+
+	// Initialize result with place and stat var dcids.
+	result := &pb.GetStatAllResponse{
+		PlaceData: make(map[string]*pb.GetStatAllResponse_PlaceStat),
+	}
+	for _, place := range places {
+		result.PlaceData[place] = &pb.GetStatAllResponse_PlaceStat{
+			StatVarData: make(map[string]*pb.ObsTimeSeries),
+		}
+		for _, statVar := range statVars {
+			result.PlaceData[place].StatVarData[statVar] = nil
+		}
+	}
+
+	// Read data from branch in-memory cache first.
+	cacheData := s.memcache.ReadParallel(
+		rowList,
+		convertToObsSeries,
+		func(rowKey string) (string, error) {
+			return keyTokens[rowKey].place + "^" + keyTokens[rowKey].statVar, nil
+		},
+	)
+
+	for token, data := range cacheData {
+		parts := strings.Split(token, "^")
+		place := parts[0]
+		statVar := parts[1]
+		if data == nil {
+			result.PlaceData[place].StatVarData[statVar] = &pb.ObsTimeSeries{}
+		} else {
+			result.PlaceData[place].StatVarData[statVar] = data.(*pb.ObsTimeSeries)
+		}
+	}
+
+	// If cache value is not found in memcache, then look up in BigTable
+	extraRowList := bigtable.RowList{}
+	for key, token := range keyTokens {
+		if result.PlaceData[token.place].StatVarData[token.statVar] == nil {
+			extraRowList = append(extraRowList, key)
+		}
+	}
+	extraData, err := readStatsPb(ctx, s.btTable, extraRowList, keyTokens)
+	for place, placeData := range extraData {
+		for statVar, data := range placeData {
+			result.PlaceData[place].StatVarData[statVar] = data
+		}
```

Comment:
should we confirm if there is anything at all to read from extraRowList before calling readStatsPb?

Reply:
done